source/common/io/io_uring_worker_impl.h

```diff
@@ -0,0 +1,120 @@
+      injected_completions_ &= ~RequestType::Connect;
+    }
+  }
+  void onRead(int32_t, bool injected) override {
+    if (injected && (injected_completions_ & RequestType::Read)) {
+      injected_completions_ &= ~RequestType::Read;
+    }
+  }
+  void onWrite(int32_t, bool injected) override {
+    if (injected && (injected_completions_ & RequestType::Write)) {
+      injected_completions_ &= ~RequestType::Write;
+    }
+  }
+
+private:
+  void unlink();
+
+  os_fd_t fd_;
+  IoUringWorkerImpl& parent_;
+  uint32_t injected_completions_{0};
+};
+
+class IoUringWorkerImpl : public IoUringWorker, protected Logger::Loggable<Logger::Id::io> {
+public:
+  IoUringWorkerImpl(uint32_t io_uring_size, bool use_submission_queue_polling,
+                    Event::Dispatcher& dispatcher);
+  IoUringWorkerImpl(std::unique_ptr<IoUring> io_uring_instance, Event::Dispatcher& dispatcher);
+  ~IoUringWorkerImpl() override;
+
+  // IoUringWorker
+  IoUringSocket& addAcceptSocket(os_fd_t fd, IoUringHandler& handler) override;
+  IoUringSocket& addServerSocket(os_fd_t fd, IoUringHandler& handler,
+                                 uint32_t read_buffer_size) override;
+  IoUringSocket& addClientSocket(os_fd_t fd, IoUringHandler& handler,
+                                 uint32_t read_buffer_size) override;
+
+  Event::Dispatcher& dispatcher() override;
+
+  Request* submitAcceptRequest(IoUringSocket& socket, sockaddr_storage* remote_addr,
+                               socklen_t* remote_addr_len) override;
+  Request* submitCancelRequest(IoUringSocket& socket, Request* request_to_cancel) override;
+  Request* submitCloseRequest(IoUringSocket& socket) override;
+  Request* submitReadRequest(IoUringSocket& socket, struct iovec* iov) override;
+  Request* submitWritevRequest(IoUringSocket& socket, struct iovec* iovecs,
+                               uint64_t num_vecs) override;
+  Request* submitConnectRequest(IoUringSocket& socket,
+                                const Network::Address::InstanceConstSharedPtr& address) override;
+
+  // From socket from the worker.
+  std::unique_ptr<IoUringSocketEntry> removeSocket(IoUringSocketEntry& socket);
+  // Inject a request completion into the io_uring instance.
+  void injectCompletion(IoUringSocket& socket, uint32_t type, int32_t result);
+  // Remove all the injected completion for the specific socket.
+  void removeInjectedCompletion(IoUringSocket& socket);
+
+protected:
+  void onFileEvent();
+  void submit();
+
+  // The io_uring instance.
+  std::unique_ptr<IoUring> io_uring_instance_;
+  // The file event of io_uring's eventfd.
+  Event::FileEventPtr file_event_{nullptr};
+  Event::Dispatcher& dispatcher_;
+  // All the sockets in this worker.
+  std::list<std::unique_ptr<IoUringSocketEntry>> sockets_;
+  // This is used to mark whether the delay submit is enabled.
+  // The IoUriingWorks delay the submit the requests which are submitted in request completion
+  // callback.
+  bool delay_submit_{false};
```

Comment:
In what cases do we want to delay the submit?

Reply:
When we process the completion events, those events callback may submit new request again, but we needn't submit those request for each of them, we can submit them all after all the events process are done. We test that, this improve the performance a lot. I will update the comment to explain this more.

Reply:
To simplify, we want to delay all these submit syscalls inside an completion event loop, so that we will have only **one** syscall at the end.